File: Mobile Application/Customer/lib/models/order.dart
// To parse this JSON data, do
//
//     final order = orderFromJson(jsonString);

import 'dart:convert';

import 'package:fuodz/constants/app_strings.dart';
import 'package:fuodz/models/delivery_address.dart';
import 'package:fuodz/models/order_stop.dart';
import 'package:fuodz/models/package_type.dart';
import 'package:fuodz/models/vendor.dart';
import 'package:fuodz/models/order_product.dart';
import 'package:fuodz/models/payment_method.dart';
import 'package:fuodz/models/user.dart';
import 'package:jiffy/jiffy.dart';

Order orderFromJson(String str) => Order.fromJson(json.decode(str));

String orderToJson(Order data) => json.encode(data.toJson());

class Order {
  Order({
    this.id,
    this.canRate,
    this.rateDriver,
    this.code,
    this.verificationCode,
    this.note,
    this.type,
    this.status,
    this.paymentStatus,
    this.subTotal,
    this.discount,
    this.deliveryFee,
    this.comission,
    this.tax,
    this.total,
    this.deliveryAddressId,
    this.paymentMethodId,
    this.vendorId,
    this.userId,
    this.driverId,
    this.createdAt,
    this.updatedAt,
    this.formattedDate,
    this.paymentLink,
    this.orderProducts,
    this.orderStops,
    this.user,
    this.driver,
    this.deliveryAddress,
    this.paymentMethod,
    this.vendor,
    //
    this.packageType,
    this.pickupLocation,
    this.dropoffLocation,
    this.pickupDate,
    this.pickupTime,
    this.recipientName,
    this.recipientPhone,
    this.width,
    this.height,
    this.length,
    this.weight,
  });

  int id;
  bool canRate;
  bool rateDriver;
  String code;
  String verificationCode;
  String note;
  String type;
  String status;
  String paymentStatus;
  double subTotal;
  double discount;
  double deliveryFee;
  double comission;
  double tax;
  double total;
  int deliveryAddressId;
  int paymentMethodId;
  int vendorId;
  int userId;
  int driverId;
  String pickupDate;
  String pickupTime;
  DateTime createdAt;
  DateTime updatedAt;
  String formattedDate;
  String paymentLink;
  List<OrderProduct> orderProducts;
  List<OrderStop> orderStops;
  User user;
  User driver;
  DeliveryAddress deliveryAddress;
  PaymentMethod paymentMethod;
  Vendor vendor;
  //Package related
  PackageType packageType;
  DeliveryAddress pickupLocation;
  DeliveryAddress dropoffLocation;
  String weight;
  String length;
  String height;
  String width;
  String recipientName;
  String recipientPhone;

  factory Order.fromJson(dynamic json) {
    return Order(
      id: json["id"] == null ? null : json["id"],
      canRate: json["can_rate"] == null ? null : json["can_rate"],
      rateDriver: json["can_rate_driver"] == null ? false : json["can_rate_driver"],
      code: json["code"] == null ? null : json["code"],
      verificationCode:
          json["verification_code"] == null ? "" : json["verification_code"],
      note: json["note"] == null ? "--" : json["note"],
      type: json["type"] == null ? null : json["type"],
      status: json["status"] == null ? null : json["status"],
      paymentStatus:
          json["payment_status"] == null ? null : json["payment_status"],
      subTotal: json["sub_total"] == null
          ? null
          : double.parse(json["sub_total"].toString()),
      discount: json["discount"] == null
          ? null
          : double.parse(json["discount"].toString()),
      deliveryFee: json["delivery_fee"] == null
          ? null
          : double.parse(json["delivery_fee"].toString()),
      comission: json["comission"] == null
          ? null
          : double.parse(json["comission"].toString()),
      //
      tax: json["tax"] == null ? null : double.parse(json["tax"].toString()),
      total:
          json["total"] == null ? null : double.parse(json["total"].toString()),
      deliveryAddressId: json["delivery_address_id"] == null
          ? null
          : int.parse(json["delivery_address_id"].toString()),
      //
      paymentMethodId: json["payment_method_id"] == null
          ? null
          : int.parse(json["payment_method_id"].toString()),
      vendorId: json["vendor_id"] == null
          ? null
          : int.parse(json["vendor_id"].toString()),
      userId: json["user_id"] == null
          ? null
          : int.parse(json["user_id"].toString()),
      driverId: json["driver_id"] == null
          ? null
          : int.parse(json["driver_id"].toString()),
      createdAt: json["created_at"] == null
          ? null
          : DateTime.parse(json["created_at"]),
      updatedAt: json["updated_at"] == null
          ? null
          : DateTime.parse(json["updated_at"]),
      formattedDate:
          json["formatted_date"] == null ? null : json["formatted_date"],
      paymentLink: json["payment_link"] == null ? "" : json["payment_link"],
      //
      orderProducts: json["products"] == null
          ? null
          : List<OrderProduct>.from(
              json["products"].map((x) => OrderProduct.fromJson(x))),
      orderStops: json["stops"] == null
          ? null
          : List<OrderStop>.from(
              json["stops"].map((x) => OrderStop.fromJson(x))),
      user: json["user"] == null ? null : User.fromJson(json["user"]),
      driver: json["driver"] == null ? null : User.fromJson(json["driver"]),
      deliveryAddress: json["delivery_address"] == null
          ? null
          : DeliveryAddress.fromJson(json["delivery_address"]),
      paymentMethod: json["payment_method"] == null
          ? null
          : PaymentMethod.fromJson(json["payment_method"]),
      vendor: json["vendor"] == null ? null : Vendor.fromJson(json["vendor"]),

      //package related data
      packageType: json["package_type"] == null
          ? null
          : PackageType.fromJson(json["package_type"]),
      pickupLocation: json["pickup_location"] == null
          ? null
          : DeliveryAddress.fromJson(json["pickup_location"]),
      dropoffLocation: json["dropoff_location"] == null
          ? null
          : DeliveryAddress.fromJson(json["dropoff_location"]),
      recipientName: json["recipient_name"],
      recipientPhone: json["recipient_phone"],
      pickupDate: json["pickup_date"] != null
          ? Jiffy(json["pickup_date"]).format("dd MMM, yyyy")
          : "",
      pickupTime: "${json["pickup_date"]} ${json["pickup_time"]}",
      // Jiffy("${json["pickup_date"]} ${json["pickup_time"]}","yyyy-MM-dd hh:mm:ss").format("hh:mm a"),
      weight: json["weight"].toString() ?? "",
      length: json["length"].toString() ?? "",
      height: json["height"].toString() ?? "",
      width: json["width"].toString() ?? "",
    );
  }

  Map<String, dynamic> toJson() => {
        "id": id,
        "code": code,
        "verification_code": verificationCode,
        "note": note,
        "type": type,
        "status": status,
        "payment_status": paymentStatus,
        "sub_total": subTotal,
        "discount": discount,
        "delivery_fee": deliveryFee,
        "comission": comission,
        "tax": tax,
        "total": total,
        "delivery_address_id": deliveryAddressId,
        "payment_method_id": paymentMethodId,
        "vendor_id": vendorId,
        "user_id": userId,
        "driver_id": driverId,
        "created_at": createdAt.toIso8601String(),
        "updated_at": updatedAt.toIso8601String(),
        "formatted_date": formattedDate,
        "payment_link": paymentLink,
        "products": List<dynamic>.from(orderProducts.map((x) => x.toJson())),
        "stops": List<dynamic>.from(orderStops.map((x) => x.toJson())),
        "user": user.toJson(),
        "driver": driver.toJson(),
        "delivery_address": deliveryAddress.toJson(),
        "payment_method": paymentMethod.toJson(),
        "vendor": vendor.toJson(),
      };

  //getters

  //
  get isPaymentPending =>
      paymentStatus == "pending" && ["pending"].contains(status);
  get isPackageDelivery => type == "package";
  get isScheduled => status == "scheduled";

  //status => 'pending','preparing','enroute','failed','cancelled','delivered'
  get canChatVendor {
    if (!AppStrings.enableChat) {
      return false;
    }
    return vendor != null &&
        ["pending", "preparing", "enroute"].contains(status);
  }

  get canChatDriver {
    if (!AppStrings.enableChat) {
      return false;
    }
    return driver != null && ["enroute"].contains(status);
  }

  bool get canCancel => ["scheduled", "pending"].contains(status);

  bool get canRateVendor {
    return this.canRate && ["cancelled", "delivered"].contains(status);
  }

  bool get canRateDriver {
    return rateDriver && driverId != null && ["cancelled", "delivered"].contains(status);
  }
}

File: app/Models/Order.php
<?php

namespace App\Models;

use App\Traits\FirebaseMessagingTrait;
use App\Traits\FirebaseDBTrait;
use Illuminate\Support\Facades\Auth;
use Illuminate\Support\Str;


class Order extends BaseModel
{

    use FirebaseMessagingTrait, FirebaseDBTrait;

    protected $fillable = ["status", "note", "reason", "type", "sub_total", "driver_id"];
    protected $with = ["user"];
    protected $appends = ["payment_link", 'formatted_date', 'formatted_type', 'can_rate', 'can_rate_driver'];


    public function scopeFullData($query)
    {
        return $query->with(["products.product", "stops.delivery_address", "user", "driver", "delivery_address", "payment_method", "vendor" => function ($query) {
            return $query->withTrashed();
        }, 'package_type', 'pickup_location', 'dropoff_location']);
    }

    public function scopeMine($query)
    {
        return $query->when(Auth::user()->hasRole('manager'), function ($query) {
            return $query->where('vendor_id', Auth::user()->vendor_id);
        });
    }

    public function products()
    {
        return $this->hasMany('App\Models\OrderProduct', 'order_id', 'id')->withTrashed();
    }

    public function stops()
    {
        return $this->hasMany('App\Models\OrderStop', 'order_id', 'id')->withTrashed();
    }

    public function user()
    {
        return $this->belongsTo('App\Models\User', 'user_id', 'id');
    }

    public function driver()
    {
        return $this->belongsTo('App\Models\User', 'driver_id', 'id');
    }

    public function delivery_address()
    {
        return $this->belongsTo('App\Models\DeliveryAddress', 'delivery_address_id', 'id');
    }

    public function payment_method()
    {
        return $this->belongsTo('App\Models\PaymentMethod', 'payment_method_id', 'id');
    }

    public function vendor()
    {
        return $this->belongsTo('App\Models\Vendor', 'vendor_id', 'id');
    }

    public function payment()
    {
        return $this->belongsTo('App\Models\Payment', 'id', 'order_id');
    }


    //
    public function package_type()
    {
        return $this->belongsTo('App\Models\PackageType', 'package_type_id', 'id');
    }

    public function pickup_location()
    {
        return $this->belongsTo('App\Models\DeliveryAddress', 'pickup_location_id', 'id')->withTrashed();
    }

    public function dropoff_location()
    {
        return $this->belongsTo('App\Models\DeliveryAddress', 'dropoff_location_id', 'id')->withTrashed();
    }





    //
    public function getCanRateAttribute()
    {

        if (empty(Auth::user())) {
            return false;
        }
        //
        $vendorReview = Review::where('user_id', Auth::id())->where('order_id', $this->id)->first();
        return empty($vendorReview);
    }

    public function getCanRateDriverAttribute()
    {

        if (empty(Auth::user())) {
            return false;
        }
        //
        $driverReview = Review::where('user_id', Auth::id())->where('driver_id', $this->driver_id)->first();
        return empty($driverReview);
    }

    public function getPaymentLinkAttribute()
    {

        if ($this->payment_status == "pending") {
            return route('order.payment', ["code" => $this->code]);
        } else {
            return "";
        }
    }

    //
    public function getFormattedTypeAttribute()
    {
        return Str::ucfirst($this->type);
    }
    public function getIsPackageAttribute()
    {
        return $this->type == "package";
    }





    //updating earning of vendor & driver
    public function updateEarning()
    {
        //'pending','preparing','ready','enroute','delivered','failed','cancelled'
        if ($this->status == 'delivered') {

            //update vendor earning
            $earning = Earning::firstOrCreate(
                ['vendor_id' => $this->vendor_id],
                ['amount' => 0]
            );

            $systemCommission = ($this->vendor->commission / 100) * $this->sub_total;
            //minus our commission 
            $earning->amount += $this->sub_total - $systemCommission;
            $earning->save();



            //update driver
            if (!empty($this->driver_id)) {

                $earning = Earning::firstOrCreate(
                    ['user_id' => $this->driver_id],
                    ['amount' => 0]
                );

                $driver = User::find($this->driver_id);
                //driver commission from delivery fee + tip from customer
                $earnedAmount = (($driver->commission / 100) * $this->delivery_fee) + $this->tip;


                //if system is using driver wallet
                $enableDriverWallet = (bool) setting('enableDriverWallet', "0");
                if ($enableDriverWallet) {

                    //
                    $driverWallet = $this->driver->wallet;
                    if (empty($driverWallet)) {
                        $driverWallet = $this->driver->updateWallet(0);
                    }

                    //if driver will be taking cash from customer, we remove the delivery fee from his/her wallet
                    if ($this->payment_method->slug == "cash") {
                        $driverWallet->balance = $driverWallet->balance - $earnedAmount;
                    }
                    //if order was paid for via card/wallet, we add the delivery fee to his/her wallet
                    else {
                        $driverWallet->balance = $driverWallet->balance + $earnedAmount;
                    }

                    //
                    $driverWallet->save();
                } else {
                    $earning->amount = $earning->amount + $earnedAmount;
                }
                $earning->save();
            }
        }
    }

    //updating wallet balance is order failed and was paid via wallet
    public function refundUser()
    {
        //'pending','preparing','ready','enroute','delivered','failed','cancelled'
        if (in_array($this->status, ['failed', 'cancelled']) && in_array($this->payment_status, ['successful'])  && $this->payment_method->slug != "cash") {

            //update user wallet
            $wallet = Wallet::firstOrCreate(
                ['user_id' => $this->user_id],
                ['balance' => 0]
            );

            //
            $wallet->balance += $this->total;
            $wallet->save();
        }
    }

    public function notifyDeliveryBoys()
    {

        if ($this->status == "ready" && !empty($this->delivery_address_id)) {
            if ($this->vendor->auto_assignment) {
                $this->saveToFCDB($this);
            }
        }
        //for parcel auto
        else if (
            $this->payment_status == "successful" &&
            in_array($this->status, ["pending", "preparing"]) &&
            !empty($this->package_type_id)
        ) {

            $packageTypePricing = PackageTypePricing::where('vendor_id', $this->vendor_id)
                ->where('package_type_id', $this->package_type_id)->first();

            if (!empty($packageTypePricing) && $packageTypePricing->auto_assignment) {
                $this->saveToFCDB($this);
            }
        }
    }

    public function clearFirestore(bool $force = false)
    {
        // if ($this->status == "enroute" || $force) {
        //     if ($this->vendor->auto_assignment) {
        //         $this->deleteFromFCDB($this);
        //     }
        // }
        if ($this->status == "enroute" || !empty($this->driver_id) || $force) {
            $this->deleteFromFCDB($this);
        }
    }
}
